src/Input/DownshiftSelect/Select.tsx

```diff
@@ -0,0 +1,355 @@
+import React, { HTMLProps, useState } from 'react';
+import { useCombobox, UseComboboxProps } from 'downshift';
+
+import * as internalComponents from './SelectStyle';
+import { ArrowDownIcon, ArrowUpIcon, CloseCircleSolidIcon } from '../..';
+import { isFunction, isUndefined, noop } from 'lodash';
+import { useId } from 'react-id-generator';
+
+export interface Item {
+  value: string;
+  label: string;
+  disabled?: boolean;
+  [x: string]: any;
+}
+
+export const defaultTransformFunction = (items: Item[], inputValue: string) =>
+  items.filter(item =>
+    item.label.toLowerCase().includes(inputValue.toLowerCase())
+  );
+
+export const itemToString = (item: Item) => (item ? item.label : '');
+
+export const defaultEmptyListText =
+  'No results found. Try another keyword to search for.';
+
+// These are our custom props for our subcomponents. We need this because
+// sometimes we want to make additional props available to the subcomponent,
+// e.g. a Item subcomponent needs to know the item it represents.
+export type ItemProps = {
+  item: Item;
+};
+
+// These are the actual types for the subcomponents. Each subcomponent takes
+// its own, custom props (defined by us) AND the props that correspond to the
+// intrinsic html element which the subcomponent "extends".
+type ContainerType = HTMLProps<HTMLDivElement>;
+type LabelType = HTMLProps<HTMLLabelElement>;
+type ComboboxType = HTMLProps<HTMLDivElement>;
+type InputType = HTMLProps<HTMLInputElement>;
+type IndicatorsContainerType = HTMLProps<HTMLDivElement>;
+type LoadingIndicatorType = HTMLProps<HTMLDivElement>;
+type ClearButtonType = HTMLProps<HTMLButtonElement>;
+type ToggleButtonType = HTMLProps<HTMLButtonElement>;
+type MenuType = HTMLProps<HTMLUListElement>;
+type ItemType = ItemProps & HTMLProps<HTMLLIElement>;
+type EmptyListType = HTMLProps<HTMLLIElement>;
+type HelperTextType = HTMLProps<HTMLSpanElement>;
+
+// This is the type for the 'components' prop on the Select. Using
+// ComponentType allows us to take in any kind of component, e.g. class
+// components, functional components or styled components.
+export interface Components {
```

Comment:
Not sure if we should have a separate file for all these types?
Then `Select.tsx` can only have the Select component.

Reply:
It's only around 30 lines of code for the types. I don't think that would make it easier.

Reply:
Oh not because of the number of lines of code. It might be a personal preference I guess. It helps in separating the responsibility of files. The types are kept in a separate file. 
Something that can be decided for standardization of the codebase later.

Reply:
I see!

Actually I don't like separating stuff by "type" so much ^^ I think it's better to organize things logically, in modules. Like we're doing on an app level, with the module folders, where we've tried to move away from the whole "reducers", "selectors", "actions" and whatnot-folders.